compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+             else: sl.b.int
+  for i in sl.a..last:
+    yield (i, n[i])
+
+proc newIdentNode(id: PIdent, info: TLineInfo): UntypedAst =
+  result.n = ast.newNodeI(nkIdent, info)
+  result.n.ident = id
+
+proc newSymNode(sym: PSym, info: TLineInfo): UntypedAst =
+  result.n = ast.newNodeI(nkSym, info)
+  result.n.sym = sym
+
+proc unreachableSymError(c: ConfigRef, s: PSym, info: TLineInfo): UntypedAst =
+  # TODO: use a dedicated report kind
+  result.n = c.newError(ast.newNodeI(nkEmpty, info),
+                        SemReport(kind: rsemUnknownIdentifier, sym: s))
+  result.hasError = true
+
+proc isReachable(c: PContext, s: PSym): bool =
+  # speed up the check by rejecting easy-to-detect not-in-scope symbols
+  # early on
+  if s.kind in skAllVars and sfGlobal notin s.flags and
+     s.itemId.module != c.module.itemId.module:
+    # the symbol is not a global and not part of the current module -> it's
+    # definitely not reachable
+    return false
+
+  # search all local scopes as well as the current module's top-level one
+  for scope in allScopes(c.currentScope):
+    let it = strTableGet(scope.symbols, s.name)
+    if it != nil and it.id == s.id: # check if it's really the same symbol
+      # it is. The symbol is in scope, so it's safe to use it
+      return true
+
+  # the symbol is not part of neither local scopes nor the top-level
+  # scope. If it's not a global, it's unreachable
+  if s.kind in skAllVars and sfGlobal notin s.flags:
+    return false
+
+  # check if the symbol is present in the hidden interface of the module it
+  # is owned by. Note that since generic instantiations are not added to
+  # the hidden interface, they'll never be treated as "in scope" here
+  for it in semtabAll(c.graph, getModule(s)).data:
+    if it != nil and it.id == s.id:
+      # found it! the symbol is reachable
+      return true
+
+  result = false
+
+proc desym(c: PContext, s: PSym, info: TLineInfo): UntypedAst =
+  ## Turns a symbol into either an identifier or a symbol node depending on
+  ## whether the latter is safe to do.
+  ## There are two aspects here: symbol visibility and symbol scope. Macros
+  ## are allowed to forego visibility rules (e.g. by inserting a symbol that
+  ## is not visible from its insertion position), but scoping rules must not be
+  ## violated, as violations would result in invalid programs. For example, a
+  ## symbol of a local defined inside a procedure must not be used outside
+  ## of said procedure.
+  ## If using `s` in the current scope would be a scoping violation, an
+  ## identifier is used
+
+  # IDEA: the new node kind ``nkSymSuggestion`` (or similar) could be
+  #       introduced. It communicates to sem that the symbol must be checked
+  #       for reachability (i.e. that the defining scope is reachable from the
+  #       usage site) first. This would allow to move the scope check out of the
+  #       sanitizer, in which case it no longer needs access to an ``PContext``
+  #       instance
```

Comment:
### Check my Understanding of the Problem

If I understand things correctly, what a macro can do to break visibility is around shadowed or overloaded symbols, some examples:

```nim
# example 1
let foo = 10
block:
  let foo = 11
  echo foo       # <-- a macro could choose to make this the outer foo
```
In the above a macro wants to emit an ast thingy, that is some path from the position of `foo` in `echo foo` to it's destination (relative to it) and/or some way to define the path from a top level/containing scope module/proc body, etc.

```nim
# example 2
proc foo(a: int) = echo "singular"

proc foo(aaaaaa: varargs[int]) = echo "plural"

foo 1 # <-- a macro could make this dispatch on the varargs and get plural
```
In the above a macro wants to emit an ast thingy, that is some way to unambiguously refer to a symbol that may or may not be overloaded.

### Solution Idea(s)

A solution for the first example might look like an `nkIdentQuery "start" "path" "parts" "to" "navigate"`. This might turn into something very akin to `jq`, `xpath`, etc... but I don't mind. To be useful the macro API likely needs a facility to get paths, eg:

- `proc getIdentPath(tree, child: NimNode): NimNode{nkIdentPath}`, and
- `proc getIdentPathFrom(tree, start, dest: NimNode): NimNode{nkIdentPath}`

In both those cases all nodes must be contained within `tree`.

A few "special" paths can be considered if we find them useful, like `"{module}"` (I just randomly chose `{}` to mean interpolate) for the containing module. Or `{proc}` for the containing proc, etc. Bonus points if we can use labels.

A solution for the the second example might be always give stable identifiers for routines that are derivable, even for a human running the algorithm by hand in at least simple cases, and then a macro author can refer to it either by formulating the identifier manually or querying for it. So perhaps it could be an `nkIdentQuery <to be determined>` as well, but it'll need a bit more of a think on the algorithm, it's basically name mangling. 😆

-----
Aside, Kotlin has a facility like this, I didn't bring it up until now because I don't want to get caught up in syntax, it might not cover all cases, but does a fair few:
- https://kotlinlang.org/spec/expressions.html#this-expressions (even though it's `this` I'm just focusing on the identifier resolution)
- https://kotlinlang.org/docs/returns.html (note the 3rd example under `Return to labels` where things get implicit labels)
- erlang is untyped, so they get away with disambiguating functions based on arity, so their imports look like `-import(somemodule, [foo/1, foo/2]).`; that's not specific enough for us I'm afraid

Reply:
Is my understanding correct in that the examples refer to AST that the macro would generate and that the problem lies in how to allow the macro to refer to the mentioned entities (which are not the ones sem would normally choose)? If yes, then the I think the current way would be `genSym`, which has the problem of producing a *symbol* instead of an unresolved *identifier*. `genSym`'ed symbols are a blocker for the sanitizer right now.

My original idea was to make `genSym` produce a unique *identifier* instead of an `sfGenSym` *symbol*, but maybe that's a solution orthogonal to `nkIdentQuery`.

If you're referring to the `IDEA` comment: that was me thinking about ways to get around the `PContext` dependency. Apologies if I'm completely misunderstanding what you mean.

Reply:
> Is my understanding correct in that the examples refer to AST that the macro would generate and that the problem lies in how to allow the macro to refer to the mentioned entities (which are not the ones sem would normally choose)? If yes, then the I think the current way would be `genSym`, which has the problem of producing a _symbol_ instead of an unresolved _identifier_. `genSym`'ed symbols are a blocker for the sanitizer right now.

Not just AST a macro generates, but I think example 1 is better solved by not having the issue in the first place. In other words, `genSym` is bad, but generating a unique identifier is what we want. At which point, we only get generated symbols for things the compiler has already figured out.

For example 2, that's a user (import, call site, etc) or macro trying to unambiguously refer to the symbol. The unique identifier doesn't solve that.

> My original idea was to make `genSym` produce a unique _identifier_ instead of an `sfGenSym` _symbol_, but maybe that's a solution orthogonal to `nkIdentQuery`.

After the unique identifier part which solves example 1 well, outside of letting the user disambiguate, but let's not bother yet. I think we're down to example 2, where query or "unambiguous fully qualified path" is required.

> If you're referring to the `IDEA` comment: that was me thinking about ways to get around the `PContext` dependency. Apologies if I'm completely misunderstanding what you mean.

It's related, but I think that's an extra bonus from solving other issues.